Failed to generate a DITA website export - cannot write to a server folder

Unable to generate a DITA website export from the Orchestrator.

# Steps to replicate

1. Ingest the November 2023 DD into Mauro.
2. Branch from this to create the next draft version
3. Sign in to the Orchestrator
4. Click on the "Branches" header link
5. Select the draft (main) branch to use
6. Click the "Publish" tab
7. Click the button to generate the Data Dictionary website

# Expected

The DITA export should complete successfully.

# Actual

The plugin failed with this exception. It looks like a hardcoded (or misconfigured) directory is used to save the export to, which does not exist.

```
2024-04-22 12:20:44,986 [080-exec-2] ERROR o.g.web.errors.GrailsExceptionResolver   : AccessDeniedException occurred when processing request: [GET] /api/nhsdd/d6741673-0f45-446c-9f18-e0171f16cfa8/publish/website
C:\Users\james. Stacktrace follows:

java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.grails.core.DefaultGrailsControllerClass$ReflectionInvoker.invoke(DefaultGrailsControllerClass.java:211)
	at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
	at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.nio.file.AccessDeniedException: C:\Users\james
	at java.base/java.nio.file.Files.createDirectory(Files.java:700)
	at java.base/java.nio.file.Files.createAndCheckIsDirectory(Files.java:807)
	at java.base/java.nio.file.Files.createDirectories(Files.java:793)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at uk.ac.ox.softeng.maurodatamapper.dita.DitaProject.getDirectory(DitaProject.groovy:251)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at uk.ac.ox.softeng.maurodatamapper.dita.DitaProject.writeToDirectory(DitaProject.groovy:121)
	at uk.nhs.digital.maurodatamapper.datadictionary.publish.website.WebsiteUtility.generateWebsite(WebsiteUtility.groovy:109)
	at uk.ac.ox.softeng.maurodatamapper.plugins.nhsdd.NhsDataDictionaryService.$tt__generateWebsite(NhsDataDictionaryService.groovy:896)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at grails.gorm.transactions.GrailsTransactionTemplate$2.doInTransaction(GrailsTransactionTemplate.groovy:94)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140)
	at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
	at uk.ac.ox.softeng.maurodatamapper.plugins.nhsdd.NhsDataDictionaryController.generateWebsite(NhsDataDictionaryController.groovy:127)
	... 13 common frames omitted
```

--- a/grails-app/services/uk/ac/ox/softeng/maurodatamapper/plugins/nhsdd/NhsDataDictionaryService.groovy
+++ b/grails-app/services/uk/ac/ox/softeng/maurodatamapper/plugins/nhsdd/NhsDataDictionaryService.groovy
@@ -100,9 +100,6 @@ class NhsDataDictionaryService {
 
     static final String NHSDD_PROPERTY_CATEGORY = 'NHS Data Dictionary'
 
-    static final String TEST_OUTPUT_PATH = "/Users/james/Desktop/ditaTest/"
-
-
     TerminologyService terminologyService
     DataModelService dataModelService
     FolderService folderService
@@ -878,7 +875,7 @@ class NhsDataDictionaryService {
         NhsDataDictionary thisDataDictionary = buildDataDictionary(thisDictionary.id)
         NhsDataDictionary previousDataDictionary = buildDataDictionary(previousVersion.id)
 
-        Path outputPath = Paths.get(TEST_OUTPUT_PATH)
+        Path outputPath = Paths.get(getTestOutputPath())
         if(!isTest) {
             outputPath = Files.createTempDirectory('changePaper')
         }
@@ -891,7 +888,7 @@ class NhsDataDictionaryService {
 
         //Path outputPath = Files.createTempDirectory('website')
 
-        Path outputPath = Paths.get(TEST_OUTPUT_PATH)
+        Path outputPath = Paths.get(getTestOutputPath())
 
         return WebsiteUtility.generateWebsite(thisDataDictionary, outputPath, publishOptions)
     }
@@ -986,4 +983,16 @@ class NhsDataDictionaryService {
         log.info('Saved [{}] model complete in {}', dataModel.label, Utils.getTimeString(endTime - startTime))
     }
 
+    /**
+     * Prepare and get a test directory name under $TEMP - replaces the use of
+     * the user desktop which might not be present on a server.
+     */
+    static String getTestOutputPath() {
+        File ditaTestDir = new File(System.getProperty("java.io.tmpdir"), "ditaTest")
+
+        if (!ditaTestDir.exists()) {
+            ditaTestDir.mkdirs()
+        }
+        ditaTestDir.absolutePath
+    }
 }

--- a/src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/publish/website/WebsiteUtility.groovy
+++ b/src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/publish/website/WebsiteUtility.groovy
@@ -119,7 +119,7 @@ class WebsiteUtility {
 
         long startTime = System.currentTimeMillis()
         ZipFile zipFile = new ZipFile(outputPath.toString() + File.separator + filename)
-        //zipFile.addFolder(new File(ditaOutputDirectory))
+        zipFile.addFolder(new File(ditaOutputDirectory))
         log.info('Zip complete in {}', Utils.timeTaken(startTime))
 
         return zipFile.getFile()

